chain/epoch_manager/src/lib.rs

```diff
@@ -288,59 +288,126 @@ impl EpochManager {
+                .unwrap_or_else(|| &ValidatorStats { expected: 0, produced: 0 })
+                .clone();
             let mut chunk_stats = ValidatorStats { produced: 0, expected: 0 };
             for (_, tracker) in chunk_validator_tracker.iter() {
                 if let Some(stat) = tracker.get(&(i as u64)) {
                     chunk_stats.expected += stat.expected;
                     chunk_stats.produced += stat.produced;
                 }
             }
-            if chunk_stats.produced * 100
-                < u64::from(chunk_producer_kickout_threshold) * chunk_stats.expected
+            _total_stake += v.stake();
+            validator_block_chunk_stats
+                .insert(account_id.clone(), BlockChunkValidatorStats { block_stats, chunk_stats });
+        }
+
+        #[allow(unused_mut)]
+        let mut exempted_validators = HashSet::<&AccountId>::new();
+        #[cfg(feature = "protocol_feature_max_kickout_stake_ratio")]
+        if checked_feature!(
+            "protocol_feature_max_kickout_stake_ratio",
+            MaxKickoutStakeRatio,
+            epoch_info.protocol_version()
+        ) {
+            let min_keep_stake = _total_stake
+                * (100_u8.checked_sub(config.validator_max_kickout_stake_ratio).unwrap_or_default()
+                    as u128)
+                / 100;
+            let mut sorted_validators = validator_block_chunk_stats
+                .iter()
+                .map(|(account, stats)| {
+                    let production_ratio =
+                        if stats.block_stats.expected == 0 && stats.chunk_stats.expected == 0 {
+                            Rational64::from_integer(1)
+                        } else if stats.block_stats.expected == 0 {
+                            Rational64::new(
+                                stats.chunk_stats.produced as i64,
+                                stats.chunk_stats.expected as i64,
+                            )
+                        } else if stats.chunk_stats.expected == 0 {
+                            Rational64::new(
+                                stats.block_stats.produced as i64,
+                                stats.block_stats.expected as i64,
+                            )
+                        } else {
+                            (Rational64::new(
+                                stats.chunk_stats.produced as i64,
+                                stats.chunk_stats.expected as i64,
+                            ) + Rational64::new(
+                                stats.block_stats.produced as i64,
+                                stats.block_stats.expected as i64,
+                            )) / 2
+                        };
+                    (account, production_ratio)
+                })
+                .collect::<Vec<_>>();
+            sorted_validators.sort_by_key(|a| a.1);
+            let mut exempted_stake: Balance = 0;
+            for (account_id, _) in sorted_validators.into_iter().rev() {
```

Comment:
More general question for my education: what determines if a validator is a bp or cop? Is it some config where the operator of the node says "I only want to be a cop?". Or is it based on stake: if you stake enough, you are a bp, otherwise you are a cop?